[FIX] Protein terminal mods in AASequence

jpfeuffer opened the pull request:
# Description
Fixes #4830 
Fixes bug in AASequence when trying to set terminal modifications.
Now correctly parses residue-specific terminal mods.
And if not found for general terminal mods, it automatically looks for protein terminal mods.
The best way in the future would be to actually check if this peptide is protein terminal and only look up the corresponding modifications but this logic needs to be outside of AASequence and implemented everywhere where we have access to this information.

# Checklist:
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing unit tests pass locally with my changes
- [x] Updated or added python bindings for changed or new classes

jpfeuffer commented:
What the hell?

jpfeuffer commented:
I did not touch anything related to XTandem. And the converted mzMLs are fine as seen on non-Windows.

jpfeuffer commented:
@cbielow Can you reproduce this?

cbielow commented:
Yes, can reproduce Comet's error.
When doing `FileConverter.exe -in C:/dev/openms_test/share/OpenMS/examples/FRACTIONS/BSA1_F1.mzML -out C:/dev/openms_test/share/OpenMS/examples/FRACTIONS/BSA1_F1.mzML`, then Comet does not crash anymore and finished successfully.. although the results are slightly different (suggesting that on Linux it also does something weird, but does not crash).

XTandem I cannot reproduce (X! TANDEM Vengeance (2015.12.15.2)).


btw, since it came up in the tests (printing `(exit code: ù).`): could you change
`(exit code: " + qp_->exitCode() + "). ` to `(exit code: " + int(qp_->exitCode()) + "). ` in
https://github.com/OpenMS/OpenMS/blob/develop/src/openms/source/SYSTEM/ExternalProcess.cpp#L117
while you're at it?
Thanks!

jpfeuffer commented:
Weird. That Fileconversion command was exactly what I did for the files. This is why they changed.

jpfeuffer commented:
I just converted again. Results in the same mzML file (except for another dataprocessing entry) and updated offsets.
Same results in all tests.
I think Windows Comet just can't read non-Windows generated things.

jpfeuffer commented:
rebuild jenkins

cbielow commented:
solves https://github.com/OpenMS/OpenMS/issues/4654 (Comet crashes due to conversion of line endings from LF to CRLF by Git, which invalidates the mzML indeX)

hroest commented on src/openms/source/CHEMISTRY/AASequence.cpp:
```diff
@@ -1429,23 +1429,53 @@ namespace OpenMS
 
-    n_term_mod_ = ModificationsDB::getInstance()->getModification(modification, "", ResidueModification::N_TERM);
+    String residue = "";
+    char last_char_no_parentheses = modification[modification.length()-2];
```
what happens if I enter a string of size 1? 

```
setNTerminalModification("T")
```

hroest replied:
@jpfeuffer can you please add that to the AASequence_test.cpp

hroest commented on src/openms/source/CHEMISTRY/AASequence.cpp:
```diff
@@ -1429,23 +1429,53 @@ namespace OpenMS
-    c_term_mod_ = ModificationsDB::getInstance()->getModification(modification, "", ResidueModification::C_TERM);
+
+    String residue = "";
+    char last_char_no_parentheses = modification[modification.length()-2];
```
same here

timosachsenberg commented on src/openms/source/CHEMISTRY/AASequence.cpp:
```diff
@@ -1445,9 +1444,19 @@ namespace OpenMS
+        n_term_mod_ = ModificationsDB::getInstance()
+            ->getModification(modification, residue, ResidueModification::N_TERM);
+      }
+      catch (...)
```
please comment try catch logic

timosachsenberg commented on src/openms/source/CHEMISTRY/AASequence.cpp:
```diff
@@ -1476,7 +1485,14 @@ namespace OpenMS
     if (!modification.hasSubstring("Protein C-term"))
     {
-      c_term_mod_ = ModificationsDB::getInstance()->getModification(modification, residue, ResidueModification::C_TERM);
+      try
```
same ehere

timosachsenberg commented:
do the pxds need an update?

jpfeuffer commented:
Unless pxds need special annotation when you catch exceptions, no. No interface changed.. yet.
It should heavily though, to make this pepXML parsing and general mod. handling from a nightmare to usable.

Failures unrelated.. I really need to check why their frequency increased after the update.

timosachsenberg commented:
I thought there were some references introduced that might be supported

cbielow commented on src/openms/source/CHEMISTRY/AASequence.cpp:
```diff
@@ -1429,23 +1428,84 @@ namespace OpenMS
+      try
+      {
+        c_term_mod_ = ModificationsDB::getInstance()->getModification(modification, residue, ResidueModification::C_TERM);
+      }
```
return from function here?

cbielow commented on src/openms/source/CHEMISTRY/AASequence.cpp:
```diff
@@ -1429,23 +1428,84 @@ namespace OpenMS
+      }
+      catch (...)
+      {
+        c_term_mod_ = ModificationsDB::getInstance()->getModification(modification, residue, ResidueModification::PROTEIN_C_TERM);
```
catch, but don't do anything.

cbielow commented on src/openms/source/CHEMISTRY/AASequence.cpp:
```diff
@@ -1429,23 +1428,84 @@ namespace OpenMS
+    }
+    else
+    {
+      c_term_mod_ = ModificationsDB::getInstance()->getModification(modification, residue, ResidueModification::PROTEIN_C_TERM);
```
and then always do this.

Would make the code easier and less repetitive?

cbielow commented on src/openms/source/FORMAT/PepXMLFile.cpp:
```diff
@@ -1319,31 +1347,71 @@ namespace OpenMS
+      entry_lower = entry_lower.toLower();
+      if (entry_lower == "y")
+      {
+        aa_mod.protein_terminus = true;
```
sorry, but all this code does not belong here!

collect all the information you need and refactor it to `aa_mod`'s implemention.
Ideally, make it a constructor and forbid changing properties individually (to maintain invariants).

This is a typical case of 'just add the code inline' which makes it
 - not testable
 - creates huge unmaintainable classes
 - is no fun to refactor at all

jpfeuffer replied:
aa_mod should not be here at all.
It is:
a) abused also for terminal mods
b) should be our own Residuemodification object, such that we do not lookup ever changing "mod-descriptions" dozens of times

cbielow commented on src/openms/source/FORMAT/PepXMLFile.cpp:
```diff
@@ -1365,6 +1433,7 @@ namespace OpenMS
       if (desc.empty())
       {
         vector<String> mods;
+        // if terminus was not specified
```
creating a string from `aa_mod` should be a member function of aa_mod!

jpfeuffer commented:
I have to give up now. Not sure when I get to it again.

cbielow commented on src/openms/source/FORMAT/PepXMLFile.cpp:
```diff
@@ -1863,7 +1863,8 @@ namespace OpenMS
       for (const auto& mod : fixed_modifications_)
       {
-        if (mod.getTerminus() == "n")
+        if (mod.getRegisteredMod()->getTermSpecificityName() == "N-term" ||
```
`.hasSuffix("N-term")`  to make it shorter?
